contracts/base/Dispatcher.sol

```diff
@@ -184,17 +184,17 @@ abstract contract Dispatcher is Payments, V2SwapRouter, V3SwapRouter, Callbacks
         internal
         returns (bool success, bytes memory output)
     {
-        (bytes memory data, address token, address spender, uint256 id) =
-            abi.decode(inputs, (bytes, address, address, uint256));
+        (bytes memory data, address token, address spender, uint256 id, address returnAddress) =
+            abi.decode(inputs, (bytes, address, address, uint256, address));
 
         // this approval is auto-wiped when the token is transferred
         ERC721(token).approve(spender, id);
 
         (success, output) = protocol.call(data);
 
-        // if !success manually revoke the approval
-        // what to do with the NFT??
-        if (!success) ERC721(token).approve(address(0), id);
+        // if the sale was not successful, return the NFT to the requested address
+        // TODO could separate this out into a sweep?
+        if (!success) ERC721(token).transferFrom(address(this), returnAddress, id);
```

Comment:
in the future it might be worth considering using reentrancy for a bulk revert here (if we end up doing that for other things).

Probably not the best for bytecode size, but this could be a public function that gets called so this entire logic can revert it it's unsuccessful, instead of catching a revert at the sudoswap call level